pkg/agent/endpoints/sdsv3/handler.go

```diff
@@ -307,6 +337,171 @@ func (h *Handler) triggerReceivedHook() {
 	}
 }
 
+type validationContextBuilder interface {
+	buildOne(resourceName, trustDomainID string) (*any.Any, error)
+	buildAll(resourceName string) (*any.Any, error)
+}
+
+type rootCABuilder struct {
+	bundles map[string]*bundleutil.Bundle
+}
+
+func newRootCABuilder(bundle *bundleutil.Bundle, federatedBundles map[spiffeid.TrustDomain]*bundleutil.Bundle) validationContextBuilder {
+	bundles := make(map[string]*bundleutil.Bundle, len(federatedBundles)+1)
+	// Add td bundle if provided
+	if bundle != nil {
+		bundles[bundle.TrustDomainID()] = bundle
+	}
+
+	for td, federatedBundle := range federatedBundles {
+		bundles[td.IDString()] = federatedBundle
+	}
+
+	return &rootCABuilder{
+		bundles: bundles,
+	}
+}
+
+func (b *rootCABuilder) buildOne(resourceName, trustDomain string) (*any.Any, error) {
+	bundle, ok := b.bundles[trustDomain]
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "no bundle found for trust domain: %q", trustDomain)
```

Comment:
I'm not sure this should be `codes.Internal`, it's easy for someone to make a request for a trust domain ID that they don't have access to. I'd recommend either `codes.InvalidRequest` or `codes.NotFound`.

Reply:
I choose codes.NotFound